File: python/Interviews/interviews/main/alphabet_discovery.py
from queue import PriorityQueue

class CustomPriorityQueue(PriorityQueue):
    class PriorityQueueEntry:
        def __init__(self, priority_queue, character):
            assert isinstance(priority_queue, CustomPriorityQueue)  #isinstance is the only way PyCharm knows the type

            self.character = character
            self.priority_queue = priority_queue

        def __lt__(self, other):
            return self.priority_queue.in_edge_count_dict[self.character] < \
                   self.priority_queue.in_edge_count_dict[other.character]

        def __repr__(self):
            return self.character

    def __init__(self, character_graph_node_dict):
        super(CustomPriorityQueue, self).__init__()

        self.in_edge_count_dict = {}
        self.character_graph_node_dict = dict(character_graph_node_dict)

        for character in self.character_graph_node_dict.keys():
            if(character not in self.in_edge_count_dict):
                self.in_edge_count_dict[character] = 0

            for antecedentCharacter in self.character_graph_node_dict.get(character).antecedents:
                if(antecedentCharacter not in self.in_edge_count_dict):
                    self.in_edge_count_dict[antecedentCharacter] = 0

                self.in_edge_count_dict[antecedentCharacter] += 1


    def remove(self, character):
        del self.in_edge_count_dict[character]
        for antecedentCharacter in self.character_graph_node_dict[character].antecedents:
            assert antecedentCharacter in self.in_edge_count_dict
            assert self.in_edge_count_dict[antecedentCharacter] > 0
            self.in_edge_count_dict[antecedentCharacter] -= 1


class AlphabetGraphNode:
    def __init__(self, character):
        self.character = character
        self.antecedents = set()

    def __repr__(self):
        return "C:" + self.character + ",A:" + str(self.antecedents)


class AlphabetDiscovery(object):
    def __init__(self, words):
        self.words = list(words)
        self.character_graph_node_dict = {}
        self.alphabet = None

    def get_alphabet(self):
        if(self.alphabet is not None):
            return self.alphabet

        if(len(self.words) is 0):
            self.alphabet = []
            return self.alphabet

        if(len(self.words) is 1):
            self.alphabet = self.words[0]
            return self.alphabet

        self.processWords(0, len(self.words), 0)

        self.alphabet = self.compute_alphabet(self.words, self.character_graph_node_dict)

        return self.alphabet

    # process words creating a directed acyclic graph
    def processWords(self, word_start_index, word_count, character_skip_count):

        # create the newly found character, and graph node
        def create_graph_node(characterSkipCount, wordIndex):
            character = self.words[wordIndex][characterSkipCount]
            if (character not in self.character_graph_node_dict):
                graph_node = AlphabetGraphNode(character)
                self.character_graph_node_dict[character] = graph_node

            return character


        word_end_index = word_start_index + word_count

        if(len(self.words[word_start_index]) == character_skip_count):
            word_start_index += 1

        current_character = create_graph_node(character_skip_count, word_start_index)

        previous_character = current_character

        flag_range_discovered = False
        range_start_index = -1
        word_index = -1
        for index in range(word_start_index + 1, word_end_index):
            word_index = index
            if(self.words[word_index][character_skip_count] == previous_character):
                if(not flag_range_discovered):
                    # start range
                    flag_range_discovered = True
                    range_start_index = word_index - 1
            else:
                # complete the range; process it
                if(flag_range_discovered):
                    assert range_start_index >= 0
                    self.processWords(range_start_index, word_index - range_start_index, character_skip_count + 1)

                    # close range
                    flag_range_discovered = False
                    range_start_index = -1

                # create the newly found character, and graph node
                current_character = create_graph_node(character_skip_count, word_index)

                # create the graph connection
                self.character_graph_node_dict[previous_character].antecedents.add(current_character)

                # reset the previous character
                previous_character = current_character

        # complete the range; process it
        if(flag_range_discovered):
            assert range_start_index >= 0
            self.processWords(range_start_index, word_index - range_start_index, character_skip_count + 1)



    def compute_alphabet(self, words, character_graph_node_dict):

        alphabet = []

        queue = CustomPriorityQueue(character_graph_node_dict)
        for character in character_graph_node_dict.keys():
            queue.put(CustomPriorityQueue.PriorityQueueEntry(queue, character))

        while(not queue.empty()):
            character_queue_entry = queue.get()
            assert queue.in_edge_count_dict[character_queue_entry.character] == 0
            queue.remove(character_queue_entry.character)
            alphabet.append(character_queue_entry.character)

        unorderedCharacters = set()

        for word in words:
            for character in word:
                if character not in self.character_graph_node_dict:
                    unorderedCharacters.add(character)

        for character in unorderedCharacters:
            alphabet.append(character)

        return alphabet

File: java/Interviews/src/main/java/interviews/graph/solution2/AlphabetDiscovery.java
package interviews.graph.solution2;

import com.google.common.base.Joiner;

import java.util.*;

/**
 * Directed Graph
 */
class AlphabetGraphNode {
  final Character character;
  final Set<Character> antecedents;

  public AlphabetGraphNode(Character character) {
    this.character = character;
    this.antecedents = new HashSet<Character>();
  }

  @Override
  public String toString() {
    return "C:" + this.character + ",A:" + Joiner.on(',').join(this.antecedents);
  }
}

public class AlphabetDiscovery {
  final List<String> words;
  String alphabet;
  final Map<Character, AlphabetGraphNode> characterGraphNodeMap;


  public AlphabetDiscovery(List<String> words) {
    this.words = words;
    this.characterGraphNodeMap = new HashMap<Character, AlphabetGraphNode>();
  }

  public String getAlphabet() {
    if(this.alphabet != null) {
      return this.alphabet;
    }

    if(this.words.size() == 0) {
      this.alphabet = "";
      return this.alphabet;
    }

    if(this.words.size() == 1) {
      this.alphabet = this.words.get(0);
      return this.alphabet;
    }

    processWords(0, this.words.size(), 0);

    this.alphabet = computeAlphabet(this.words, this.characterGraphNodeMap, new CharacterComparator());

    return this.alphabet;
  }

  // process words creating a directed acyclic graph
  void processWords(int wordStartIndex, int wordCount, int characterSkipCount) {
    int wordIndex = wordStartIndex;
    int wordEndIndex = wordStartIndex + wordCount;

    if(this.words.get(wordStartIndex).length() == characterSkipCount) {
      wordIndex++;
    }

    // create the newly found character, and graph node
    Character currentCharacter = this.words.get(wordIndex).charAt(characterSkipCount);
    if(!this.characterGraphNodeMap.containsKey(currentCharacter)) {
      this.characterGraphNodeMap.put(currentCharacter, new AlphabetGraphNode(currentCharacter));
    }

    Character previousCharacter = currentCharacter;
    boolean flagRangeDiscovered = false;
    int rangeStartIndex = -1;
    for(wordIndex++; wordIndex < wordEndIndex; wordIndex++) {
      if(this.words.get(wordIndex).charAt(characterSkipCount) == previousCharacter) {
        if(!flagRangeDiscovered) {
          // start range
          flagRangeDiscovered = true;
          rangeStartIndex = wordIndex - 1;
        }

      } else {

        // complete the range; process it
        if(flagRangeDiscovered) {
          assert rangeStartIndex >= 0;
          processWords(rangeStartIndex, wordIndex - rangeStartIndex, characterSkipCount + 1);

          // close range
          flagRangeDiscovered = false;
          rangeStartIndex = -1;
        }

        // create the newly found character, and graph node
        currentCharacter = this.words.get(wordIndex).charAt(characterSkipCount);
        if(!this.characterGraphNodeMap.containsKey(currentCharacter)) {
          this.characterGraphNodeMap.put(currentCharacter, new AlphabetGraphNode(currentCharacter));
        }

        // create the graph connection
        this.characterGraphNodeMap.get(previousCharacter).antecedents.add(currentCharacter);

        // reset the previous character
        previousCharacter = currentCharacter;
      }
    }

    // complete the range; process it
    if(flagRangeDiscovered) {
      assert rangeStartIndex >= 0;
      processWords(rangeStartIndex, wordIndex - rangeStartIndex, characterSkipCount + 1);
    }

  } // for

  class CharacterComparator implements Comparator<Character> {
    final Map<Character, Integer> inEdgeCountMap;

    public CharacterComparator() {
      this.inEdgeCountMap = new HashMap<Character, Integer>();
      for(Character character : characterGraphNodeMap.keySet()) {
        if(!this.inEdgeCountMap.containsKey(character)) {
          this.inEdgeCountMap.put(character, 0);
        }

        for(Character antecedentCharacter : characterGraphNodeMap.get(character).antecedents) {
          if(!this.inEdgeCountMap.containsKey(antecedentCharacter)) {
            this.inEdgeCountMap.put(antecedentCharacter, 0);
          }

          this.inEdgeCountMap.put(antecedentCharacter, this.inEdgeCountMap.get(antecedentCharacter) + 1);
        }
      }
    }

    @Override
    public int compare(Character firstCharacter, Character secondCharacter) {
      return Integer.compare(this.inEdgeCountMap.get(firstCharacter), this.inEdgeCountMap.get(secondCharacter));
    }

    void removeCharacter(Character character) {
      this.inEdgeCountMap.remove(character);
      for(Character antecedentCharacter : characterGraphNodeMap.get(character).antecedents) {
        assert this.inEdgeCountMap.containsKey(antecedentCharacter);
        assert this.inEdgeCountMap.get(antecedentCharacter) > 0;
        this.inEdgeCountMap.put(antecedentCharacter, this.inEdgeCountMap.get(antecedentCharacter) - 1);
      }
    }
  }

  static String computeAlphabet(
      final List<String> words,
      final Map<Character, AlphabetGraphNode> characterGraphNodeMap,
      final CharacterComparator characterComparator) {

    StringBuilder alphabet = new StringBuilder();

    PriorityQueue<Character> characterPriorityQueue = new PriorityQueue<Character>(characterComparator);
    characterPriorityQueue.addAll(characterGraphNodeMap.keySet());
    while(!characterPriorityQueue.isEmpty()) {
      Character character = characterPriorityQueue.remove();
      assert characterComparator.inEdgeCountMap.get(character) == 0;
      characterComparator.removeCharacter(character);
      alphabet.append(character);
    }

    Set<Character> unorderedCharacters = new HashSet<Character>();

    for(String word : words) {
      for(int characterIndex = 0; characterIndex < word.length(); characterIndex++) {
        Character character = word.charAt(characterIndex);
        if(!characterGraphNodeMap.containsKey(character)) {
          unorderedCharacters.add(character);
        }
      }
    }

    alphabet.append(Joiner.on("").join(unorderedCharacters));

    return alphabet.toString();
  }

}
